Add customization point for printing code involving Derivative instances

bjodah opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
- gh-26663
- https://github.com/sympy/sympy/issues/11802#issuecomment-602855301

#### Brief description of what is fixed or changed
Currently `lambdify` cannot properly handle `Derivative` instances. This PR allows printers to add methods for printing `Derivative` instances, which then allows lambdify to work with those seamlessly. If the special printing method `_print_Derivative_####` is not present, the exception thrown will indicate this in its error message.

#### Other comments
If this approach is approved, then I will:

- [ ] Describe `_print_Derivative_####` in the documentation for our printer classes.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* printing
  * Subclasses of CodePrinter can now customize printing of `Derivative` instances
    by implementing e.g. `_print_Derivative_zeta(args, deriv_orders)`. This allows lambdify
    to work with expressions involving unexpanded `Derivative` instances.

<!-- END RELEASE NOTES -->

bjodah committed:
Add customization point for printing code involving Derivative instances

See gh-11802 and gh-26663 for background and motivation.

bjodah committed:
Fix gh-26690

bjodah committed:
Add another test exercising _print_Derivative_FUNCNAME

bjodah committed:
Support Derivative instances as arguments in lambdify (with cse=True)

bjodah committed:
tidy up

oscarbenjamin commented on sympy/printing/codeprinter.py:
```diff
@@ -462,6 +462,16 @@ def _print_Function(self, expr):
+            return self._print_not_supported(expr, extra_info=f"\nPrinter {self.__class__.__name__} has no method: {meth_name}")
+        orders = dict(wrt_order_pairs)
+        seq_orders = [orders[arg] for arg in obj.args]
+        return pmeth(obj.args, seq_orders)
```
This assumes that `arg` will be in the orders mapping I think. What happens for e.g. `Derivative(zeta(x**2), x)`?

bjodah replied:
That's a good question. Is that a well formed input? Should we throw when the inner arguments are `Expr` instead of `Symbol`?

Currently `.diff()` gives an expression with `Subs`:
```python
>>> zeta(x**2).diff(x)
2*x*Subs(Derivative(zeta(_xi_1), _xi_1), _xi_1, x**2)
```
and now I discover that this isn't handled by `lambdify`, it would perhaps be useful if we could make that work here as well:
```python
>>> f = lambdift(x, zeta(x**2).diff(x), modules=['mpmath'])
>>> f(2)
...
NameError: name 'Subs' is not defined
```

bjodah replied:
`Subs` are now hopefully supported by `lambdify` and `CodePrinter._print_Derivative` will now require that the arguments of its inner function are all `Symbol`.

bjodah committed:
Support zeta(x**2).diff(x) in lambdify

bjodah committed:
Avoid setting 'strict': False in internal printer used in lambdify.

Fix according to @oscarbenjamin's comment:
https://github.com/sympy/sympy/issues/26690#issuecomment-2163750478

bjodah committed:
_is_safe_ident only works for str

bjodah committed:
Merge branch 'master' into CodePrinter-customization-point-Derivative

bjodah commented:
I don't quite understand the failing mypy test:
```
sympy/printing/tensorflow.py:106: error: Incompatible types in assignment (expression has type "Callable[[Any], Any]", base class "CodePrinter" defined the type as "Callable[[Any, Any], Any]")
````
That line assigns to `_print_Relational`, which I'm not changing in this PR:
https://github.com/sympy/sympy/blob/8255cf243fbaf5142a44f93ad6a5bff62b4de6cf/sympy/printing/tensorflow.py#L106

oscarbenjamin commented:
> That line assigns to `_print_Relational`, which I'm not changing in this PR:

In the superclass this is assigned to `_print_not_supported` whose signature you have changed in this PR:
https://github.com/sympy/sympy/blob/8255cf243fbaf5142a44f93ad6a5bff62b4de6cf/sympy/printing/codeprinter.py#L614

bjodah commented:
Ah, I see, thanks. But I'm not convinced this is a problem, should we silence mypy here using some magic comment?

oscarbenjamin commented:
We can add `# type: ignore`.

I think it is a reasonable complaint though because the signature of all the functions like `print_Relational` is supposed to be that they only take a single positional argument. Here you have implicitly changed the signature of 20 different methods and mypy is observing that these signatures are not compatible with those in the subclass.

bjodah committed:
Restore signature of CodePrinter._print_not_supported

oscarbenjamin commented:
Really we want the fix for gh-26690 to be backported to the 1.13 release branch. It's awkward to do that if you bundle different changes together like this with commits that undo other commits and so on.

Is there anything more to it than the one line change from https://github.com/sympy/sympy/issues/26690#issuecomment-2163750478?

bjodah commented:
Sorry, I don't quite follow: this PR predates gh-26690, if you just need a commit for back-porting (which is what I gather from your last comment), then I've now opened gh-26713, but the fix is yours, not mine. But I do agree that your fix was better than mine, that's why I included it in [1845439](https://github.com/sympy/sympy/pull/26678/commits/18454392624e0eca2d5c34992b10a0bb7efe8b16) in this PR.

As for adding commits that undo other commits, are you referring to [656a582](https://github.com/sympy/sympy/pull/26678/commits/656a5826bc063e1e536b966ade336ad0725b8c83)? Because that is me addressing your review (which I appreciate, and agree with, by the way).

I thought we tried not to rebase during the review phase (except just before merging, when appropriate), just because it is confusing when github drops earlier comments, etc. But I can keep rebasing during the process as well if you insist?

oscarbenjamin commented:
What you are doing with the commits here is fine. I just wanted to be able to cherry-pick the fix for gh-26690 because as far as I know it is the only unresolved regression reported for 1.13rc1 i.e. currently the only release blocker:
https://github.com/sympy/sympy/milestone/65

While I suggested the fix I didn't test it extensively and I don't know the lambdify code very well.

When more changes get bundled together it makes cherry-picking or backporting more complicated. This is mainly not for technical reasons but because it is more complicated to think about what is being changed, whether there is any risk of new regressions or if any of the other changes is more controversial than the simple bugfix for the known regression.

Technically though cherry-picking (with `git cherry-pick`) does not really work if the change that is wanted is in a commit that is undoing something from another commit as is the case for 18454392624e0eca2d5c34992b10a0bb7efe8b16. The diff is wrt to a commit that includes changes I don't want to cherry-pick rather than wrt master and then the test is in a different commit as well.

In some projects (e.g. CPython) there is a strong emphasis on keeping the purpose of each PR separate and then squashing all commits at the point of merge (gh-18548). Under that model cherry picking and backporting work more smoothly and it is routine to merge a PR to main and then backport to 3.13, 3.12, 3.11, ... with the backports being usually automated.

bjodah commented:
I see. Thank you for giving some details. I failed to realize this was the only release blocker.

bjodah committed:
Merge branch 'master' into CodePrinter-customization-point-Derivative

oscarbenjamin commented:
I think a merge/rebase is needed to fix the authors check.

bjodah committed:
Merge branch 'master' into CodePrinter-customization-point-Derivative

oscarbenjamin commented:
Okay, looks good. Thanks.
